app/modules/tunnel/socket_service.py

```
@@ -0,0 +1,253 @@
+            finally:
+                await pubsub.unsubscribe(channel)
+                await pubsub.close()
+
+        try:
+            return await asyncio.wait_for(_wait_one_response(), timeout=timeout + 2.0)
+        except asyncio.TimeoutError:
+            from app.modules.tunnel.tunnel_service import TunnelConnectionError
+            raise TunnelConnectionError(
+                f"Tool call timed out for workspace_id={workspace_id} (endpoint={endpoint})",
+                last_error="timeout",
+            )
```

Ruff (0.15.1)

[warning] Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)